### PR TITLE
Nested Destructuring

### PR DESCRIPTION
**Pull request by gdotdesign**
This implements a recursive algorithm for destructuring and unifies statement and case branch destructuring checks. This allows:

- Destructuring with values in them
- Mixing and nesting destucturings
- Using complex destructurings in `if` expressions

Some examples:

```mint
case Test::Record(name: Maybe::Just({"John", "Doe"}), age: 32) {
  Test::Record(Maybe::Just({"John", a})) => a
  => "WTF"
}
    
case Result::Ok(Maybe::Just({"Hello", "Bello"})) {
  Result::Ok(Maybe::Just({"Hello", b})) => "Matches"
  Result::Ok(Maybe::Just({"JULIE"})) => "Not matches"
  Result::Err => "Error"
  => "Else branch"
}

case ["a", "b", "d"] {
  [a, ...b, d] => b
  ["a"] => ["one"]
  [a] => ["something else"]
  [] => ["empty"]
}

if Result::Ok(Maybe::Just({"Hello", b})) = result {
  ...
} else {
  ...
}
```

Top level exhaustiveness checking is still there but if any branch has nested destructuring or value (`"string"` for example) in it then it's excluded from the check.

Follow-up work (in separate PRs):
- Move `_match` and related functions into the runtime
- Allow named destructuring of records (including in enums) 
- At some point implement exhaustiveness checking in sublevels

### CODE REVIEW

File: core/source/Http.mint [hunk 1]
```diff
@@ -334,7 +334,7 @@ module Http {
             (new DOMParser()).parseFromString(responseText, "text/html");
 
           const errorNode =
-            doc.querySelector("parsererror");
+            object.querySelector("parsererror");
```
Comment (gdotdesign) [inline core/source/Http.mint:337]: This is an unrelated fix which will probably go into a different commit.

File: spec/compilers/case [hunk 2]
```diff
@@ -1,36 +1,38 @@
 component Main {
-  fun test : Bool {
+  fun render : String {
```
Comment (Sija) [inline spec/compilers/case:2]:
```suggestion
  fun render : Bool {
```

File: src/parsers/variable.cr [hunk 3]
```diff
@@ -45,7 +45,7 @@ module Mint
       variable(track) || raise error
     end
 
-    def variable(track = true, invalid = [] of String) : Ast::Variable?
+    def variable(track = true, invalid = ["true", "false"]) : Ast::Variable?
```
Comment (Sija) [inline src/parsers/variable.cr:48]: I'd keep the constant for performance reasons.

```suggestion
    def variable(track = true, invalid = INVALID_VARIABLE_NAMES) : Ast::Variable?
```

**Comment by gdotdesign** [comment]
> @gdotdesign Does it resolves #283?

Yes it will :)

File: spec/compilers/case (hunk 2 above)
Comment (gdotdesign) [reply spec/compilers/case:2]: String is the return type of this function from the value below the `case`.

File: src/parsers/variable.cr (hunk 3 above)
Comment (gdotdesign) [reply src/parsers/variable.cr]: Moved it from the arguments into the method body.

File: src/compilers/top_level.cr [hunk 4]
```diff
@@ -266,6 +266,113 @@ module Mint
         const _S = mint.Store
         const _E = mint.Enum
 
+        const _PR = (patterns) => new RecordPattern(patterns)
+        const _PE = (x, pattern) => new Pattern(x, pattern)
+        const _PV = Symbol("Variable");
+        const _PS = Symbol("Sperad");
```
Comment (Sija) [inline src/compilers/top_level.cr:272]:
```suggestion
        const _PV = Symbol("Variable")
        const _PS = Symbol("Spread")
```

File: src/parsers/variable.cr [hunk 5]
```diff
@@ -45,15 +45,15 @@ module Mint
...
+    def variable(track = true) : Ast::Variable?
       start do |start_position|
         value = gather do
           next unless char.ascii_lowercase?
           letters_or_numbers
         end
 
         next unless value
-        next if value.in?(invalid)
+        next if value.in?(%w[true false])
```
Comment (Sija) [inline src/parsers/variable.cr:56]: Please, move it into a constant to avoid redundant allocations.

File: src/type_checkers/case.cr [hunk 6]
```diff
@@ -24,7 +24,7 @@ module Mint
 
       unified =
         node
-          .branches[1..node.branches.size]
+          .branches[(1..)]
```
Comment (Sija) [inline src/type_checkers/case.cr:27]:
```suggestion
          .branches[1..]
```

File: src/type_checkers/case.cr [hunk 7]
```diff
@@ -84,7 +89,15 @@ module Mint
           } if !not_matched.empty? && !catch_all
         elsif condition.name == "Array"
           destructurings =
-            node.branches.map(&.match).select(Ast::ArrayDestructuring)
+            node.branches
+              .map(&.match)
+              .select(Ast::ArrayDestructuring)
+              .select do |branch|
```
Comment (Sija) [inline src/type_checkers/case.cr:95]:
```suggestion
              .select! do |branch|
```

File: src/type_checkers/case.cr [hunk 8]
```diff
@@ -50,19 +50,24 @@ module Mint
...
-        if item
+        if parent
           not_matched =
-            item.options.reject do |option|
+            parent.options.reject do |option|
               node
                 .branches
                 .compact_map(&.match)
                 .any? do |match|
                   case match
```
Comment (Sija) [inline src/type_checkers/case.cr:63]:
```suggestion
              node
                .branches
                .any? do |branch|
                  case match = branch.match
```
